### PR TITLE
Startup screen code reorg

### PR DESCRIPTION
**Pull request by pyrollo**
This PR is an attempt to reorganize start up screen code (main menu and loading screen).

I wanted to continue Kilbith's #9973 but I quickly got lost in code mess and duplication.

This PR gathers everything to manage startup screen except main menu in one single class. This class is will be responsible of drawing background (clouds or texture), header, footer and progress information (during loading).

Formerly, this was scattered in various files : guiEngine, RenderingEngine and Clientlauncher. Many other where using access to RenderingEngine singleton to update loading message. They all get now the instance of StartupScreen.

## To do

This PR is a Work in Progress

- There are still bugs to be fixed but mainly works (not tested all menu things).
- MenuTextureSource should be replaced by a StartupTextureSource;

In a (near or far) future: 
- Main menu could be reworked to offer possibility to plainly customize screen background and loading screen;
- Other scenes could be added instead of clouds;

## How to test

If you want, you can just compile, launch and observe main menu and loading screen. It should be close to what's already there.

Side effect (looks like an improvement): When launching a game, background is kept same as in main menu. So, game title and background are now displayed while loading.

### CODE REVIEW

File: src/client/startup_screen.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	bool             tile;
+	unsigned int     minsize;
+} image_definition;
+
+class StartupScreen {
+public:
+	StartupScreen();
+	~StartupScreen();
+
+	enum backgroundType {
```
Comment (rubenwardy) [inline src/client/startup_screen.h:40]: Same for other enums

```suggestion
	enum class BackgroundType {
```

File: src/client/startup_screen.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+#include <string>
+
+class Clouds;
+class MenuTextureSource;
+namespace irr{namespace scene{class ISceneManager;}}
+
+class StartupScreen;
+extern StartupScreen *g_startup_screen;
+
+typedef struct {
```
Comment (rubenwardy) [inline src/client/startup_screen.h:29]: Spot the C dev :) 

```c++
struct ImageDefinition {
	video::ITexture *texture = nullptr;
	bool             tile;
	unsigned int     minsize;
};
```

File: src/client/startup_screen.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+*/
+
+#include <string>
+
+class Clouds;
+class MenuTextureSource;
+namespace irr{namespace scene{class ISceneManager;}}
+
+class StartupScreen;
+extern StartupScreen *g_startup_screen;
```
Comment (rubenwardy) [inline src/client/startup_screen.h:27]: ewwww, why

This should be passed as dependencies into using classes instead

File: src/client/startup_screen.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	unsigned int     minsize;
+} image_definition;
+
+class StartupScreen {
+public:
+	StartupScreen();
+	~StartupScreen();
+
+	enum backgroundType {
+		BT_COLOR,
```
Comment (rubenwardy) [inline src/client/startup_screen.h:41]:
```suggestion
		COLOR,
```

File: src/client/startup_screen.h (hunk 2 above)
Comment (pyrollo) [reply src/client/startup_screen.h:29]: :) Actually most of the code is copy paste from older code. That's a good chance to clean up things.

File: src/client/startup_screen.h (hunk 3 above)
Comment (pyrollo) [reply src/client/startup_screen.h:27]: This global basically replaced g_menuclouds and g_menucloudsmgr (I guess still a C dev signature).

Five classes uses StartupScreen : `guiEngine`, `Client`, `ClientLauncher`, `Game` and `ModApiMainMenu`, may be to complicated for passing instance.

Some sort of static getStartupScreen() returning a singleton would avoid both global variable and passing instance to each class.

File: src/client/startup_screen.h (hunk 1 above)
Comment (pyrollo) [reply src/client/startup_screen.h]: Not so simple change but done.

**Comment by rubenwardy** [comment]
Closed due to inactivity

**Pull request closed without merging**